Implement UDAFs on datafusion

mwylde opened the pull request:
Adds proper support for UDAFs in datafusion by implementing an Accumulator which collects values for each UDAF argument, then on evaluate passes them into the UDAF dylib function. The state is the collectesd vector of records.

jacksonrnewhouse commented on crates/arroyo-df/src/udafs.rs:
```diff
@@ -0,0 +1,228 @@
+    }
+
+    fn size(&self) -> usize {
+        std::mem::size_of_val(self)
```
This is supposed to be the total size of the allocator, i.e. include the sizes of the arrays it owns.

jacksonrnewhouse reviewed (commented):
It looks like in parsing the UDAF code we allow a mixture of vectors and scalars, but then we throw in execution. Is that right, and are we happy with how we treat each case?

mwylde commented:
Addressed the comments.

With regards to the mix scalar/aggregate args, I would prefer this be a plan time failure but I'm not sure how best to accomplish that, since datafusion seems comfortable mixing argument types. Do you think it's possible for us to implement that check in a visitor?

Otherwise, we could support calling udafs with scalar args by making an array out of the scalar, but that probably isn't what the user wants and it would be better to error.
